### PR TITLE
Add allow_uturn parameter to match api

### PR DESCRIPTION
**Pull request by kaligrafy**
Related to #3595

### CODE REVIEW

File: include/engine/api/match_parameters.hpp [hunk 1]
```diff
@@ -61,12 +62,16 @@ struct MatchParameters : public RouteParameters
     }
 
     template <typename... Args>
-    MatchParameters(std::vector<unsigned> timestamps_, Args... args_)
-        : RouteParameters{std::forward<Args>(args_)...}, timestamps{std::move(timestamps_)}
+    MatchParameters(
+        std::vector<unsigned> timestamps_,
+        const boost::optional<bool> allow_uturn_,
+        Args... args_)
+        : RouteParameters{std::forward<Args>(args_)...}, timestamps{std::move(timestamps_), allow_uturn{allow_uturn_}}
```
Comment (daniel-j-h) [inline include/engine/api/match_parameters.hpp]: This needs to be done in a backwards compatible way: this is libosrm's public interface which we can't change without bumping the major version number.

You could add a constructor overload taking the new parameter and keeping the old one as is. Then add a note (and once we merged this PR open a ticket) for removing this backwards compatible workaround in v6. And tag it with the milestone v6.

File: include/engine/api/match_parameters.hpp [hunk 2]
```diff
@@ -61,12 +62,16 @@ struct MatchParameters : public RouteParameters
...
     {
     }
 
     std::vector<unsigned> timestamps;
+    boost::optional<bool> allow_uturn = true;
```
Comment (daniel-j-h) [inline include/engine/api/match_parameters.hpp]: Should this be an optional or a boolean with true being the default? Why the optional here?

File: include/server/api/match_parameter_grammar.hpp [hunk 3]
```diff
@@ -32,6 +32,11 @@ struct MatchParametersGrammar final : public RouteParametersGrammar<Iterator, Si
             qi::lit("timestamps=") >
             (qi::uint_ %
              ';')[ph::bind(&engine::api::MatchParameters::timestamps, qi::_r1) = qi::_1];
+        
+        allow_uturn_rule =
+            qi::lit("allow_uturn=") >
+            (qi::uint_ %
+             ';')[ph::bind(&engine::api::MatchParameters::allow_uturn, qi::_r1) = qi::_1];
```
Comment (daniel-j-h) [inline include/server/api/match_parameter_grammar.hpp]: This needs a changelog entry, docs/http.md entry and more unit tests.

**Comment by TheMarex** [comment]
Could we make this consistent with the `route` plugin? We already have a parameter called `continue_straight` which enables overriding the default u-turn behavior.

https://github.com/Project-OSRM/osrm-backend/blob/master/docs/http.md#route-service

The behavior should be as follows:
- `continue_straight` not specified by the user: auto detection of u-turns in the trace by map matching (current behavior)
- `continue_straight=false` same behavior as above
- `continue_straight=true` (aka `allow_uturns=false`) disable uturns detection in map matching plugin

**Comment by kaligrafy** [comment]
If I understand correctly, the continue_straight parameter is already available in the match parameters since it is inherited, isn't it? So we just need to send it to the filterCandidates method?

**Comment by daniel-j-h** [comment]
@kaligrafy Yes.

**Comment by kaligrafy** [comment]
I don't know why, but the continue_straight is not passed from the route parameters parser. I guest it is because they are not include in the base rule in osrm-backend/include/server/api/route_parameters_grammar.hpp? Can somedy just tell me how to make sure it propagates to the match parameters parser?

**Comment by oxidase** [comment]
@kaligrafy yes, it is just a `route` parameter. small change should fix it 
[a.txt](https://github.com/Project-OSRM/osrm-backend/files/730223/a.txt)

**Comment by daniel-j-h** [comment]
The match grammar uses the route grammar as base for parsing

https://github.com/Project-OSRM/osrm-backend/blob/445225bf6dc719ef4f91fe6ab7e6fa2f4941fef7/include/server/api/match_parameter_grammar.hpp#L25

and the route grammar parses continue straight

https://github.com/Project-OSRM/osrm-backend/blob/445225bf6dc719ef4f91fe6ab7e6fa2f4941fef7/include/server/api/route_parameters_grammar.hpp#L34-L37

You can access continue straight via `parameters.continue_straight` here:

https://github.com/Project-OSRM/osrm-backend/blob/445225bf6dc719ef4f91fe6ab7e6fa2f4941fef7/src/engine/plugins/match.cpp#L116

I just tried it locally on master and this compiles just fine.

(Note: continue straight is a [optional](https://github.com/Project-OSRM/osrm-backend/blob/445225bf6dc719ef4f91fe6ab7e6fa2f4941fef7/include/engine/api/route_parameters.hpp#L107) you can check `if (opt) use_boolean(*opt)`)

**Comment by kaligrafy** [comment]
It compiles but using it in a request triggers a query error.

Envoyé de mon iPhone

**Comment by daniel-j-h** [comment]
Yes sorry for that: use [the patch](https://github.com/Project-OSRM/osrm-backend/files/730223/a.txt) @oxidase provided for now. We opened a ticket to keep parameters and grammars in sync in the future: https://github.com/Project-OSRM/osrm-backend/issues/3614

**Comment by TheMarex** [comment]
I think it is fine if you copy the `continue_straight` rule from the route plugin to the match plugin.

**Comment by kaligrafy** [comment]
I'll check that as soon as possible.

**Comment by daniel-j-h** [comment]
What's the status here? Do you need help running this out @kaligrafy? Or where are you blocked?

**Comment by kaligrafy** [comment]
No, I just have to finish a big project at the job and I have no time left for this right now. I'll check as soon as possible!

**Comment by daniel-j-h** [comment]
👋 is this still on your todo list?

If not feel free to close this pull request, the issue already captures the task.

**Comment by daniel-j-h** [comment]
Any updates here or can this be closed?

**Comment by kaligrafy** [comment]
Yes, I prefer that somebody fix this since I do not have the time to do so right now.

Thanks!

**Pull request closed without merging**